### DISCUSSION TITLE
Reason, technologies, comparison, goals

### DISCUSSION THREAD

**Discussion by gsal** [rio-labs/rio#33] [General]
Hello, developer team: 

I just run into this rather young project; reading the docs does not tell me much about it.  Can you guys add to the docs things like: 

- reasons for the project
- what it is you don't like about existing pure-python web development frameworks like Solara, Nicegui, Dash, Shiny, Reflex, etc.
- how you are doing differently and/or overcoming what you see as other project's short-comings
- what technologies you are using behind-the-scenes...react, vue, pywebview, webassembly, webpack, etc.

**Comment by Sn3llius** [comment]
Thanks for asking! We're currently putting together a detailed blog post about our tech stack and what makes Rio unique. I hope to have it ready by the end of next week, and I'll share it here once it's done. 😊In the meantime, here's a brief overview: 

### Reason for the project
At work, we used Streamlit, Dash, and Reflex, while at university, we worked with Shiny. Each of these frameworks has its own limitations and unique styles for developing applications. This led us to create Rio, a brand new GUI framework designed to enable the creation of modern web apps with just a few lines of Python.

Our goal is to simplify web and app development by adhering to Python's core principles of simplicity and compactness. With Rio, there's no need to learn additional languages like HTML, CSS, or JavaScript, as all UI, logic, components, and even layout are handled entirely in Python. We’ve eliminated the distinction between front-end and back-end, with Rio managing all communication transparently.

We aim to create a framework that allows for rapid prototyping and the building of complex nested UI layouts using a single tool. By being as Pythonic as possible, Rio ensures a rapid learning curve and provides users with the flexibility to achieve whatever they desire. Additionally, we focus on making Rio intuitive and powerful, enabling developers to transition seamlessly from idea to implementation.

### What we didn't like about existing pure Python web development frameworks
Our key features highlight the main aspects we found lacking in the frameworks you mentioned. It's the combination of these elements that sets us apart and significantly improves the user experience.

- **Full-Stack Web Development:** Rio handles front-end and backend for you. In fact, you won't even notice they exist. Create your UI, and Rio will take care of the rest.
- **Python Native:** Rio apps are written in 100% Python, meaning you don't need to write a single line of CSS or JavaScript.
- **Modern Python:** We embrace modern Python features, such as type annotations and asynchrony. This keeps your code clean and maintainable, and helps your code editor help you out with code completions and type checking.
- **Python Debugger Compatible:** Since Rio runs on Python, you can connect directly to the running process with a debugger. This makes it easy to identify and fix bugs in your code.
- **Declarative Interface:** Rio apps are built using reusable components, inspired by react, flutter & vue. They're declaratively combined to create modular and maintainable UIs.
- **Dynamic Attribute Binding:** Rio enables dynamic attribute binding between components without the need for event handlers. This simplifies data flow and makes your code more concise.
- **DevTool Features:** Rio's comprehensive development toolkit includes code and live reloading for seamless updates, in-app error reporting to maintain focus, and component validation for enhanced code integrity, streamlining the development process.
- **Batteries included:** Over 50 builtin components based on Google's Material Design

### What sets us apart
This question delves into our philosophy. We've dedicated significant effort to make Rio feel truly Pythonic. Rio transperantly handles  front-end and backend for you, making their existence practically invisible to the user. We've designed an intuitive layout system that requires no knowledge of HTML or CSS, streamlining the development process. Our state management is robust, allowing for global, session, and component-level states, and it automatically tracks state changes. Debugging is straightforward, with compatibility features that simplify the process when needed. Rio also comes equipped with a variety of development tools, comparable to Dash.

While Rio has its limitations as a young project, we have ambitious plans for the future. We aim to significantly enhance customization options and boost performance. We are confident that we will achieve substantial progress in the coming months.

### Used technologies
We leveraged FastAPI to manage API requests and used Uvicorn for serving our web applications. On the front end, our stack included TypeScript, CSS, and HTML, ensuring a responsive and dynamic user experience. Websockets were employed to enable real-time client-server communication. For the backend, Python was the primary language, handling the core logic and also functioning as the interface for user interactions. Additionally, we plan to incorporate Rust for performance optimizations.
